cpp/arcticdb/async/async_store.hpp

```diff
@@ -404,65 +404,47 @@ class AsyncStore : public Store {
-                                          auto de_dup_key =
-                                              de_dup_map ? de_dup_map->get_key_if_present(key_seg.atom_key())
-                                                         : std::nullopt;
-
-                                          if (de_dup_key) {
-                                              ARCTICDB_DEBUG(log::version(),
-                                                             "Found existing key with same contents: using existing object {}",
-                                                             de_dup_key.value());
-                                              (*res)[key_seg.id()] = (de_dup_key.value());
-                                              return std::make_pair(storage::KeySegmentPair{}, false);
-                                          } else {
-                                              ARCTICDB_DEBUG(log::version(),
-                                                             "No existing key with same contents: writing new object {}",
-                                                             key_seg.atom_key());
-                                              (*res)[key_seg.id()] = (key_seg.atom_key());
-                                              return std::make_pair(std::move(key_seg), true);
-                                          }
-                                      }).thenValue([key_segs = key_segs, key_seg_mutex = key_seg_mutex](auto &&pk) {
-                    auto pair_key_write = std::forward<decltype(pk)>(pk);
-                    if (pair_key_write.second) {
-                        std::lock_guard kl{*key_seg_mutex};
-                        ARCTICDB_DEBUG(log::version(), "Enqueuing {} for write", pair_key_write.first.atom_key());
-                        key_segs->emplace_back(std::move(pair_key_write.first));
-                    }
-                }));
+        auto encode_futs = folly::window(key_segments, [*this] (auto&& ks) {
+            auto [key, seg] = std::move(ks);
+            return async::submit_cpu_task(
+                EncodeAtomTask(std::move(key),
+                               ClockType::nanos_since_epoch(),
+                               std::move(seg),
+                               codec_,
+                               0, //TODO pointless, get rid
+                               encoding_version_));
+            }, write_count);
+
+            for(auto&& encode_fut : encode_futs) {
+                futs.emplace_back(
+                    std::move(encode_fut).thenValue([de_dup_map](auto &&ks) -> std::pair<VariantKey,
+                                                                                         std::optional<Segment>> {
+                            auto key_seg = std::forward<decltype(ks)>(ks);
+                            if (auto de_dup_key = de_dup_map->get_key_if_present(key_seg.atom_key())) {
```

Comment:
Previously:

```
auto de_dup_key = de_dup_map ? de_dup_map->get_key_if_present(key_seg.atom_key()) : std::nullopt;
```

Now:

`auto de_dup_key = de_dup_map->get_key_if_present(key_seg.atom_key())`

That's a behavioural change, no? What if `de_dup_map` is not set?